### DISCUSSION TITLE
FREERTOS.h supports AVR, there is an rp2040_mbed library

### DISCUSSION THREAD

**Discussion by CameronCode22** [pschatzmann/arduino-audio-tools#1705] [Q&A] [answered]
Hi, I am trying to play audio from my Arduino nano rp2040 running on mbed_os architecture. I found your library. I found the .h file named "src/AudioI2S/I2SRP2040-MBED.h"

I thought that maybe this would be compatible with the board. I am compiling in arduino IDE. I get the following error:

fatal error: FreeRTOS.h: No such file or directory
 #  include "FreeRTOS.h"
 
 I did at one point download the FreeRTOS. But I could see it was only supporting AVR architecture, as mentioned this is mbed_os. I was compiling the simple star wars example in your readme file. Is there another way or a different type of RTOS I should be using for my board?

**Comment by pschatzmann** [comment, marked as answer]
I committed some corrections but I recommend to use the more popular core for Earle Phil Hower instead

**Reply by CameronCode22** [reply]
I found another library that you created here:

[https://github.com/pschatzmann/rp2040-i2s](url)

I have ran into another issue. And I am not entirely sure if it belongs here. But you may be able to guide some sort of direction. I am using the Cyberon voice recognition model with arduino rp2040 nano. MBED_os architecture. It has an inbuilt mic. When I combine both working voice recognition and sd card playback using I2S. It hangs in the begin I2S function. Code attached below.

so in the setup it prints "I2S trying" and then gets stuck and prints nothing else. Orange light on nano board. I fear there may be some resource conflict between the model and I2S. But I have fairly basic understanding. Would be grateful for a nudge in the right direction. Thanks,

Slight edit, its worth noting. The I2S works fine on its own with the SD card set-up. It seems to be in combination with the Cyberon model. It gets caught in this hang state.

``
#include <Arduino.h>
#include <DSpotterSDK_MakerHL.h>
#include <LED_Control.h>

#include <I2S.h>
#include <SD.h>
#include <SPI.h>

// The DSpotter License Data.
#include "CybLicense.h"
#define DSPOTTER_LICENSE g_lpdwLicense

// The DSpotter Keyword Model Data.
#if defined(TARGET_ARDUINO_NANO33BLE) || defined(TARGET_PORTENTA_H7) || defined(TARGET_NICLA_VISION)
// For ARDUINO_NANO33BLE and PORTENTA_H7
#include "Model_L1.h"             // The packed level one model file.
// For NANO_RP2040_CONNECT
#elif defined(TARGET_NANO_RP2040_CONNECT)
#include "Model_L0.h"             // The packed level zero model file.
#endif
#define DSPOTTER_MODEL g_lpdwModel

// The VR engine object. Only can exist one, otherwise not worked.
static DSpotterSDKHL g_oDSpotterSDKHL;

const int chipSelect = 29; // SD card chip select pin
File wavFile;

// Callback function for VR engine
void VRCallback(int nFlag, int nID, int nScore, int nSG, int nEnergy)
{
  if (nFlag==DSpotterSDKHL::InitSuccess)
  {
      //ToDo
  }
  else if (nFlag==DSpotterSDKHL::GetResult)
  {
      /*
      When getting an recognition result,
      the following index and scores are also return to the VRCallback function:
          nID        The result command id
          nScore     nScore is used to evaluate how good or bad the result is.
                     The higher the score, the more similar the voice and the result command are.
          nSG        nSG is the gap between the voice and non-command (Silence/Garbage) models.
                     The higher the score, the less similar the voice and non-command (Silence/Garbage) models are.
          nEnergy    nEnergy is the voice energy level.
                     The higher the score, the louder the voice.
      */
      //ToDo
  }
  else if (nFlag==DSpotterSDKHL::ChangeStage)
  {
      switch(nID)
      {
          case DSpotterSDKHL::TriggerStage:
            LED_RGB_Off();
            LED_BUILTIN_Off();
            break;
          case DSpotterSDKHL::CommandStage:
            LED_BUILTIN_On();
            break;
          default:
            break;
      }
  }
  else if (nFlag==DSpotterSDKHL::GetError)
  {
      if (nID == DSpotterSDKHL::LicenseFailed)
      {
          //Serial.print("DSpotter license failed! The serial number of your device is ");
          //Serial.println(DSpotterSDKHL::GetSerialNumber());
      }
      g_oDSpotterSDKHL.Release();
      while(1);//hang loop
  }
  else if (nFlag == DSpotterSDKHL::LostRecordFrame)
  {
      //ToDo
  }
}

void setup()
{
  // Init LED control
  LED_Init_All();

  // Init Serial output for show debug info
  Serial.begin(9600);
  while(!Serial);
  DSpotterSDKHL::ShowDebugInfo(true);

  // Init VR engine & Audio
  if (g_oDSpotterSDKHL.Init(DSPOTTER_LICENSE, sizeof(DSPOTTER_LICENSE), DSPOTTER_MODEL, VRCallback) != DSpotterSDKHL::Success)
    return;
  
  // Initialize SD card
  if (!SD.begin(chipSelect)) {
    Serial.println("Failed to initialize SD card!");
  } else {
    Serial.println("Initialised SD card");
  }

    // Open the WAV file
  wavFile = SD.open("lords.wav");
  if (!wavFile) {
    Serial.println("Failed to open WAV file!");
  } else {
    Serial.println("Opened WAV file");
  }

  // Skip WAV header (44 bytes for standard WAV files)
  wavFile.seek(44);

  // Start I2S with a sample rate of 44100 Hz (CD quality), 16 bits per sample
  Serial.println("trying I2S");
  if (!I2S.begin(22500)) {
    Serial.println("Failed to initialize I2S!");
  unsigned long startTime = millis(); // Record the start time
  unsigned long threshold = 1000;    // Timeout duration in milliseconds (1 second)

  while (1) {
    if (millis() - startTime > threshold) { // Check if the timeout duration has elapsed
      Serial.println("Timeout reached!");
      break; // Exit the loop if the timeout has been reached
    }
    // Optional: Add code to retry initialization or perform other checks
    delay(100); // Small delay to avoid overwhelming the processor
  }
} else {
  Serial.println("I2S initialized successfully");
}
}

void loop()
{
  // Do VR
  g_oDSpotterSDKHL.DoVR();
}
